### PR TITLE
Add dictionary encoded block

### PR DESCRIPTION
**Pull request by nileema**

### CODE REVIEW

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDictionaryStreamReader.java [hunk 1]
```diff
@@ -160,17 +158,11 @@ public void readBatch(Object vector)
...
-                sliceVector.vector[i] = null;
-            }
-            else if (inDictionary[i]) {
-                sliceVector.vector[i] = dictionary[dataVector[i]];
-            }
-            else {
-                sliceVector.vector[i] = rowGroupDictionary[dataVector[i]];
-            }
-        }
+        int[] ids = Arrays.copyOfRange(dataVector, 0, nextBatchSize);
```
Comment (dain) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDictionaryStreamReader.java]: can we allocate the vectors to the correct size to begin with

File: presto-spi/src/main/java/com/facebook/presto/spi/block/DictionaryBlockEncoding.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+            // ignore if compacting a dictionary is not supported
+        }
+
+        // positionCount
+        int positionCount = dictionaryBlock.getPositionCount();
+        sliceOutput.appendInt(positionCount);
+
+        // dictionary
+        Block dictionary = dictionaryBlock.getDictionary();
+        dictionary.getEncoding().writeBlock(sliceOutput, dictionary);
```
Comment (dain) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/DictionaryBlockEncoding.java]: I'd use `dictionaryEncoding` here

File: presto-spi/src/main/java/com/facebook/presto/spi/block/DictionaryBlockEncoding.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    public void writeBlock(SliceOutput sliceOutput, Block block)
+    {
+        // The down casts here are safe because it is the block itself the provides this encoding implementation.
+        DictionaryBlock dictionaryBlock = (DictionaryBlock) block;
+
+        try {
+            dictionaryBlock = dictionaryBlock.compact();
+        }
+        catch (UnsupportedOperationException e) {
+            // ignore if compacting a dictionary is not supported
```
Comment (dain) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/DictionaryBlockEncoding.java]: When does this happen?

**Comment by dain** [comment]
looks good

File: presto-spi/src/main/java/com/facebook/presto/spi/block/DictionaryBlockEncoding.java (hunk 3 above)
Comment (nileema) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/DictionaryBlockEncoding.java]: It happens if the values block of the dictionary block does not implement ``copyPositions`` (array block, interleaved block)

Comment (nileema) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/DictionaryBlockEncoding.java]: I should handle this in the ``compact`` code and return the same dictionary. I'll change it.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDictionaryStreamReader.java (hunk 1 above)
Comment (nileema) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDictionaryStreamReader.java]: We know do not know the batchSize when we create the readers, we compute it just before we start reading.

**Pull request merged**